### ISSUE TITLE
include BCC in encoded output based on configuration flag

### ISSUE THREAD

**Issue by grossadamm** [mikel/mail#864]
I'm trying to get the BCC included in the RFC 2822 format output for use with the gmail api. 

I completely understand the reasoning for not including the bcc by default in Issue #426, however, I still need it to be present so I can let gmail know as I'm not sending the individual emails out myself. What is the best way for me to go about doing so. 

I'm perfectly content code diving and adding the functionality that will add the bcc address to the encoded output only if some config flag is set. And I would of course create a corresponding pull request, but I don't want to reinvent the wheel if something similar already exists.

**Comment by bf4** [comment]
You can monkey patch in the meantime to something like (untested code)

```ruby
class Mail::BccField
  def self.supress_encoding=(supress_encoding)
    @supress_encoding = supress_encoding
  end
  def self.supress_encoding
    @supress_encoding || self.supress_encoding = true
  end
  def encode
     if self.class.supress_encoding
       ''
     else
       do_encode(CAPITALIZED_FIELD)
    end
  end
end
```

and then `Mail::BccField.supress_encoding = false`

**Comment by grossadamm** [comment]
Ok, so I'm guessing a way doesn't exist. I'll attempt to monkey-patch for now. Thanks!

**Comment by bf4** [comment]
Would yo mind elaborating on where in your workflow with gmail the bug, with respect to gmail needing a bcc field, shows up?

**Comment by grossadamm** [comment]
For the gmail api, to send an email you HTTP POST to the api an rfc2822 format email encoded in base64. [Api reference](https://developers.google.com/gmail/api/v1/reference/users/messages/send). The api supports the bcc field, however when I'm generating the rfc2822 format email via this gem the bcc field is not included. 

I understand the reason why the bcc field is being stripped out in the normal case, but in my arguably abnormal case, the stripping of the field makes it impossible for me to bcc messages using this gem to generate the rfc2822 format.

**Comment by Exoth** [comment]
Yep, I need such an option to. Just an argument for to_s would be great for me.

**Comment by grossadamm** [comment]
@bf4 I assume you are a busy person and accept you probably won't have the time to look at this let alone merge it for some time.

But whenever you get a chance, could you please review my pull request and see if there's anything you would change?

Thanks for the help!

**Comment by bf4** [comment]
Cc @jeremy

**Comment by jeremy** [comment]
BCC is used to add destinations to the envelope, not the message. This gets kinda confusing because we do parse the field from messages that have it, but we also use it for delivery. They should be separate concepts. Make our BCC API add envelope destinations, then have your gmail delivery agent copy those from the envelope to the message BCC field. (If you're using Action Mailer, the `delivering_email` hook is where you'd intercept and gmailify the outbound email.)

One route, add a flag to the bcc field so you can explicitly mark it for encoding: `message.bcc.include_in_headers = true`.

Another route, stop using the message BCC field as the envelope BCC address list: change `Message#bcc=` to keep a separate `@envelope_bcc_destinations` list. The gmail delivery agent would "just" copy the envelope BCC destinations over to a message BCC field, then.

**Comment by grossadamm** [comment]
@jeremy Thanks for your time and explanation! 

My gmail delivery is much simpler than I think you imagine. Instead of actually attempting to send an email using a delivery handler (via `deliver!` or otherwise) and hooking into ActionMailer, I simply HTTP POST to their api the results of encoding the message. The generation of the POST body looks similar to this oversimplified block.

    def to_google_email_format
      {
        message: {
          raw: Mail.new(
            from: user.email,
            to: to,
            cc: cc,
            bcc: bcc,
            subject: subject,
            body: body
          ).to_s
        }
      }.to_json
    end

Generating the full envelope and then pulling the addresses back out again seems a bit roundabout for this scenario, unless I'm being completely daft.

I do like your method of just flagging the individual bcc field on the message. I believe, this is essentially what I'm already doing in the pull request, but by flagging the bcc field instance on the message instead of the bcc field class it would resolve @Exoth's issue. I can go ahead and make that change.

 I _think_ I understand what you are getting at with the separate `@envelope_bcc_destinations` list. And I may be completely daft, but the second solution seems dependent on the idea that the envelope is being created for delivery, which isn't my case (or what I can gather @Exoth's case either).

**TL;DR** Thank you for explaining, I'll modify my pull request to flag just that instance of the bcc field on the message for encoding and update the corresponding tests.

**Comment by grossadamm** [comment]
Pull request modified

**Comment by jeremy** [comment]
:heart: #872
